### PR TITLE
crosscluster/logical: check UDT equivalency during LDR creation

### LINKED ISSUE #132167: crosscluster/logical: validate UDT equivalency in LDR planning
When #131130 merges, we will support replication of a table with a  UDT if the user in is in yolo mode, i.e. with the `SKIP SCHEMA CHECK` option. To allow UDT support in general, we need to validate UDT equivalency during LDR planning via the `CheckLogicalReplicationCompatibility` function, [here](https://github.com/azhu-crl/cockroach/blob/destroy-reader-tenant/pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go#L261). 

With #131130, the source types will be plumbed to the `streampb.ReplicationProducerSpec` [here](https://github.com/azhu-crl/cockroach/blob/destroy-reader-tenant/pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go#L222).

At first, we should check physical udt equivalency, i.e. `color: blue=1, green =2` does not equal `color: green=1, blue =2` for now. 

Jira issue: CRDB-42861

Epic CRDB-41465

### PR DESCRIPTION
**Pull request by rafiss**
This check requires that the logical and physical representations of each type are identical. In the future, we may investigate ways to only require logical equivalency.

fixes https://github.com/cockroachdb/cockroach/issues/132167
Release note (ops change): When creating a logical replication stream,
any user-defined types in the source and destination are now checked for
equivalency. This allows for creating a stream that handles user-defined
types without needing to use the `WITH SKIP SCHEMA CHECK` option as long
as the replication stream uses `mode = immediate`.

### CODE REVIEW

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 1]
```diff
@@ -2100,14 +2102,14 @@ func TestLogicalReplicationCreationChecks(t *testing.T) {
...
+	// Verify that the stream cannot be created with mismatched user defined types.
 	dbA.Exec(t, "DROP TRIGGER my_trigger ON tab")
 	dbA.Exec(t, "CREATE TYPE mytype AS ENUM ('a', 'b', 'c')")
-	dbB.Exec(t, "CREATE TYPE b.mytype AS ENUM ('a', 'b', 'c')")
+	dbB.Exec(t, "CREATE TYPE b.mytype AS ENUM ('a', 'b')")
 	dbA.Exec(t, "ALTER TABLE tab ADD COLUMN enum_col mytype NOT NULL")
 	dbB.Exec(t, "ALTER TABLE b.tab ADD COLUMN enum_col b.mytype NOT NULL")
 	dbA.ExpectErr(t,
-		`cannot create logical replication stream: destination table tab column enum_col has user-defined type USER DEFINED ENUM: public.mytype`,
+		`cannot create logical replication stream: .* destination type USER DEFINED ENUM: public.mytype has logical representations \[a b c\], but the source type USER DEFINED ENUM: mytype has \[a b\]`,
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/logical_replication_job_test.go:2112]: could you also add a test case for composite type mismatch?

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go [hunk 2]
```diff
@@ -208,14 +209,26 @@ func createLogicalReplicationStreamPlanHook(
 			return err
 		}
 
+		sourceTypes := make([]*descpb.TypeDescriptor, len(spec.TypeDescriptors))
+		for i, desc := range spec.TypeDescriptors {
+			sourceTypes[i] = &desc
+		}
+		// TODO(rafi): do we need a different type resolver?
```
Comment (msbutler) [inline pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:216]: if the user is using the sql path, we still need to fast fail here until #132164 is resolved.

**Review by msbutler — APPROVED** [review]
Amazing! Also, could you modify the commit message/release note to explain that this enables UDT support on the mode=immediate?  I assume you're following up with another pr to fix https://github.com/cockroachdb/cockroach/issues/132164

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go [hunk 3]
```diff
@@ -1925,20 +1925,22 @@ func TestUserDefinedTypes(t *testing.T) {
 	// Create the same user-defined type both tables.
 	dbA.Exec(t, "CREATE TYPE my_enum AS ENUM ('one', 'two', 'three')")
 	dbB.Exec(t, "CREATE TYPE my_enum AS ENUM ('one', 'two', 'three')")
+	dbA.Exec(t, "CREATE TYPE my_composite AS (a INT, b TEXT)")
+	dbB.Exec(t, "CREATE TYPE my_composite AS (a INT, b TEXT)")
 
-	dbA.Exec(t, "CREATE TABLE data (pk INT PRIMARY KEY, val my_enum DEFAULT 'two')")
-	dbB.Exec(t, "CREATE TABLE data (pk INT PRIMARY KEY, val my_enum DEFAULT 'two')")
+	dbA.Exec(t, "CREATE TABLE data (pk INT PRIMARY KEY, val1 my_enum DEFAULT 'two', val2 my_composite)")
```
[the remark this answers is not in the document]
Comment (rafiss) [reply pkg/ccl/crosscluster/logical/logical_replication_job_test.go:1931]: 😃

File: pkg/ccl/crosscluster/logical/logical_replication_job_test.go (hunk 1 above)
Comment (rafiss) [reply pkg/ccl/crosscluster/logical/logical_replication_job_test.go:2112]: done

File: pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go (hunk 2 above)
Comment (rafiss) [reply pkg/ccl/crosscluster/logical/create_logical_replication_stmt.go:216]: done

**Review by msbutler — APPROVED** [review]

**Comment by rafiss** [comment]
tftr!

bors r+

**Pull request merged**